asserts/gpgkeypairmgr.go

```diff
@@ -67,39 +68,115 @@ func (gkm *gpgKeypairManager) gpg(input []byte, args ...string) ([]byte, error)
 	if err != nil {
 		return nil, err
 	}
 	if len(out) == 0 {
-		return nil, fmt.Errorf("cannot find key %q in GPG keyring", keyID)
+		return nil, fmt.Errorf("cannot retrieve key with fingerprint %q in GPG keyring", fpr)
 	}
 
 	pubKeyBuf := bytes.NewBuffer(out)
-	privKey, err := newExtPGPPrivateKey(pubKeyBuf, "GPG", gkm.sign)
+	privKey, err := newExtPGPPrivateKey(pubKeyBuf, "GPG", func(content []byte) ([]byte, error) {
+		return gkm.sign(fpr, content)
+	})
 	if err != nil {
-		return nil, fmt.Errorf("cannot use GPG key %q: %v", keyID, err)
+		return nil, fmt.Errorf("cannot load GPG public key with fingerprint %q: %v", fpr, err)
 	}
-	gotID := privKey.PublicKey().ID()
-	if gotID != keyID {
-		return nil, fmt.Errorf("got wrong key from GPG, expected %q: %s", keyID, gotID)
+	gotFingerprint := privKey.fingerprint()
+	if gotFingerprint != fpr {
+		return nil, fmt.Errorf("got wrong public key from GPG, expected fingerprint %q: %s", fpr, gotFingerprint)
 	}
 	return privKey, nil
 }
 
-func (gkm *gpgKeypairManager) sign(fingerprint string, content []byte) ([]byte, error) {
+// Walk iterates over all the RSA private keys in the local GPG setup calling the provided callback until this returns true for done, or errors.
+func (gkm *GPGKeypairManager) Walk(consider func(privk PrivateKey, fingerprint string) (done bool, err error)) error {
+	// see GPG source doc/DETAILS
+	out, err := gkm.gpg(nil, "--batch", "--list-secret-keys", "--fingerprint", "--with-colons")
+	if err != nil {
+		return err
+	}
+	lines := strings.Split(string(out), "\n")
+	n := len(lines)
+	if n > 0 && lines[n-1] == "" {
+		n--
+	}
+	if n == 0 {
+		return nil
+	}
+	lines = lines[:n]
+	for j := 0; j < n; j++ {
+		line := lines[j]
+		if !strings.HasPrefix(line, "sec:") {
+			continue
+		}
+		secFields := strings.Split(line, ":")
+		if len(secFields) < 5 {
+			continue
+		}
+		if secFields[3] != "1" { // not RSA
+			continue
+		}
+		keyID := secFields[4]
+		if j+1 >= n || !strings.HasPrefix(lines[j+1], "fpr:") {
+			continue
+		}
+		fprFields := strings.Split(lines[j+1], ":")
+		if len(fprFields) < 10 {
+			continue
+		}
+		fpr := fprFields[9]
+		if !strings.HasSuffix(fpr, keyID) {
+			continue // strange, skip
+		}
+		privKey, err := gkm.retrieve(fpr)
+		if err != nil {
+			return err
+		}
+		done, err := consider(privKey, fpr)
+		if err != nil {
+			return err
+		}
+		if done {
+			return nil
+		}
+	}
+	return nil
+}
+
+func (gkm *GPGKeypairManager) Put(authorityID string, privKey PrivateKey) error {
+	// NOTE: we don't need this initially at least and this keypair mgr is not for general arbitrary usage
+	return fmt.Errorf("cannot import private key into GPG keyring")
+}
+
+func (gkm *GPGKeypairManager) Get(authorityID, keyHash string) (PrivateKey, error) {
+	var hit PrivateKey
+	match := func(privk PrivateKey, fpr string) (bool, error) {
+		if privk.PublicKey().SHA3_384() == keyHash {
+			hit = privk
+			return true, nil
```

Comment:
Instead of done+err, where err is never used, we can use the error to define the stop condition.
```
stop := fmt.Errorf("stop marker")
...
if ... {
    return stop
}
...
if err == stop {
        return hit, nil
}
if err != nil {
        return nil, err
}
return nil, ... not found ..